employee-rostering-frontend/src/ui/components/DataTable.tsx

```diff
@@ -0,0 +1,178 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import {
+  Table,
+  TableHeader,
+  TableBody,
+  headerCol,
+  IRow
+} from '@patternfly/react-table';
+import {Button, ButtonVariant} from '@patternfly/react-core';
+import { SaveIcon, CloseIcon, EditIcon, TrashIcon } from '@patternfly/react-icons';
+import { EditableComponent } from './EditableComponent';
+
+export interface DataTableProps<T> {
+  title: string;
+  columnTitles: string[];
+  tableData: T[];
+}
+
+interface DataTableState<T> {
+  newRow: IRow | null;
+  currentFilter: (rowData: T) => boolean;
+}
+
+export abstract class DataTable<T,C, P extends DataTableProps<T>> extends React.Component<P, DataTableState<T>> {
+  constructor(props: P) {
+    super(props);
+    this.createNewRow = this.createNewRow.bind(this);
+    this.cancelAddingRow = this.cancelAddingRow.bind(this);
+    this.convertDataToTableRow = this.convertDataToTableRow.bind(this);
+    this.state = {newRow: null, currentFilter: (t) => true};
+  }
+
+  abstract displayDataRow(data: T): JSX.Element[];
+  abstract editDataRow(dataStore: C, data: T): JSX.Element[];
+  abstract createNewDataRow(dataStore: C): JSX.Element[];
+  abstract isValid(editedValue: C): boolean;
+  abstract extractDataFromRow(oldValue: T|{}, editedValue: C): T;
+
+  abstract updateData(data: T): void;
+  abstract addData(data: T): void;
+  abstract removeData(data: T): void;
+
+  createNewRow() {
+    if (this.state.newRow === null) {
+      const dataStore: any = {};
+      const newRowComponents: JSX.Element[] = this.createNewDataRow(dataStore);
+      const newRow: IRow = { cells: newRowComponents.map(c => {return {title: c}}).concat([{
+        title: this.getAddButtons(dataStore)}])};
+      this.setState({...this.state, newRow: newRow});
+    }
+  }
+
+  cancelAddingRow() {
+    this.setState({...this.state, newRow: null});
+  }
+
+  getAddButtons(dataStore: C): JSX.Element {
+    return <span>
+      <Button aria-label="Save"
+        variant={ButtonVariant.link}
+        onClick={() => {
+          if (this.isValid(dataStore)) {
+            this.addData(this.extractDataFromRow({},dataStore));
+            this.cancelAddingRow();
+          }
+        }}>
+        <SaveIcon />
+      </Button>
+      <Button aria-label="Cancel"
+        variant={ButtonVariant.link}
+        onClick={this.cancelAddingRow}>
+        <CloseIcon />
+      </Button>
+    </span>;
+  }
+
+  getEditButtons(dataStore: C, data: T, editableComponents: EditableComponent[]): JSX.Element {
+    return <EditableComponent ref={(c) => {editableComponents[editableComponents.length-1] = c as EditableComponent;}}
+      viewer={<span>
+        <Button aria-label="Edit"
+          variant={ButtonVariant.link}
+          onClick={() => {
+            editableComponents.forEach(c => c.startEditing());
```

Comment:
This is one of the places that make the code complicated. Edit/Save/Cancel buttons do not need to know about other editable components. Buttons shouldn't interact with other components directly. Instead, they should be given a callback function that is handled by their parent. This callback function can for example update the `state` with ID of the row whose Edit/Cancel button was clicked. EditableComponent can receive `isEditing` as a prop which can be derived from DataTable's state containing the (list of) edited row IDs.